hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -1641,7 +1641,7 @@ void optimizeARM(Vunit& unit, const Abi& abi, bool regalloc) {
   simplify(unit);
 
   lowerForARM(unit);
-
+  reuseImmq(unit);
```

Comment:
Ok I understand, just three remarks:

1) with your approach you don't get immediates, that are shared, cannot be embedded into an (e.g. logical) instruction and build up within one mov* instruction (e.g. `ldimmq{0x2300, %0} ; andqi{%0, %1, %2}` -> `andqi{0x2300}` -> `movz x1, 0x2300 ;  and x0, x0, x1`). There would be optimizing potential for them as well.

2) your implementation of isMultiword() is too eager, as it does not consider all variants of mov* (movz 16, movz immr/imms, movk, movn). So you would reuse immediates, that would be embedded in an (e.g. logical) instruction anyways (e.g. `ldimmq{0xbfffffffffffffff, %0} ; andqi{%0, %1, %2}` -> `andqi{0xbfffffffffffffff}` -> `and x0, x0, 0xbfffffffffffffff`). This might end up in more code being generated (in the case when you emit an additional `add` to calculate the target immediate). When you put your pass after foldImms(), then you would not see that effect.

3) Regarding the `ldimml{}` (from the comments below): I agree that it is not used very often directly (only irlower-intrinsic.cpp uses that instruction directly). However in vasm-arm.cpp we are lowering `addli{}`, `andli{}`, `cmpli{}`, `subli{}`, `testli{}` and `storeli{}` to `ldimml{}`s. And these are emitted more often. So there would be optimizing potential for them as well (e.g. andli{} could share a non-embeddable mask (e.g. 0x2300)).